feat: Replacing jsonwebtoken with fast-jwt

radomird opened the pull request:
Replacing the `jsonwebtoken` with more performant `fast-jwt` library.

Since options for `fast-jwt` are somewhat different, I will update the options exposed through `fastify-jwt`.

**Important note**: This is a **BREAKING CHANGE**, so after this PR is approved and merged, a new major version of the plugin will be released.

Closes #174 
Closes #166 

**Update 10-11-2021**: `fast-jwt` has been updated to support password protected private keys, This PR includes those changes as well.

<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: https://github.com/fastify/fastify/blob/master/CONTRIBUTING.md

By making a contribution to this project, I certify that:

* (a) The contribution was created in whole or in part by me and I
  have the right to submit it under the open source license
  indicated in the file; or

* (b) The contribution is based upon previous work that, to the best
  of my knowledge, is covered under an appropriate open source
  license and I have the right under that license to submit that
  work with modifications, whether created in whole or in part
  by me, under the same open source license (unless I am
  permitted to submit under a different license), as indicated
  in the file; or

* (c) The contribution was provided directly to me by some other
  person who certified (a), (b) or (c) and I have not modified
  it.

* (d) I understand and agree that this project and the contribution
  are public and that a record of the contribution (including all
  personal information I submit with it, including my sign-off) is
  maintained indefinitely and may be redistributed consistent with
  this project or the open source license(s) involved.
-->

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

darkgl0w commented on test/test.test.js:
```diff
@@ -2,16 +2,17 @@
 const helper = require('./helper')
 
-const passphrase = 'super secret passphrase'
+// TODO: Uncomment when (if) fast-jwt adds suport for encoded private keys (passphrase protected)
+// const passphrase = 'super secret passphrase'
+// const { privateKey: privateKeyProtected, publicKey: publicKeyProtected } = helper.generateKeyPairProtected(passphrase)
+// const { privateKey: privateKeyProtectedECDSA, publicKey: publicKeyProtectedECDSA } = helper.generateKeyPairECDSAProtected(passphrase)
```
Now that https://github.com/nearform/fast-jwt/pull/117 has been merged we can uncomment this.

radomird replied:
Done

darkgl0w commented on test/test.test.js:
```diff
@@ -1126,6 +817,316 @@ test('sign and verify with RSA/ECDSA certificates and global options', function
         })
     })
   })
+  // TODO: Uncomment when (if) fast-jwt adds suport for encoded private keys (passphrase protected)
```
Now that nearform/fast-jwt#117 has been merged we can uncomment this.

radomird replied:
Done

darkgl0w reviewed (approved):
LGTM !

Some suggestions:
- We should take the opportunity offered by this PR to rework the file structure of the plugin and move `jwt.test-d.ts` to `test/types/jwt.test-d.ts`, rename `test/test.test.js` to `test/jwt.test.js` and update the corresponding references in `package.json` file.
- Update to require at least `fastify >= 3.0.0` (currently it require `fastify >= 3.0.0-alpha-1`).
- We should add an explicit warning in the `README.md` stating that we migrated from `jsonwebtoken` to `fast-jwt` and that it may break "exotic" implementations even though it is a 1:1 feature implementation.

NB: this PR superseeds and closes https://github.com/fastify/fastify-jwt/pull/166 too.

radomird commented:
A new version of `fast-jwt` has been released. I've updated it and also addressed the issues mentioned in the PR. It's ready for reviews.

The following proposal from @darkgl0w :
> - We should take the opportunity offered by this PR to rework the file structure of the plugin and move jwt.test-d.ts to test/types/jwt.test-d.ts, rename test/test.test.js to test/jwt.test.js and update the corresponding references in package.json file.

falls out of scope of this PR so I'll create a separate ticket for that.

darkgl0w reviewed (approved):
LGTM with a little addition to the `README.md` note.

I spotted this breaking use case in the wild, but I guess this PR will land in a semver major.
It's not a blocking case to me as I always use values expressed in milliseconds but this plugin is widely used so WDYT ? Is it worth adding this feature to `fast-jwt` before landing this PR ?

radomird commented on README.md:
```diff
@@ -5,7 +5,9 @@
-JWT utils for Fastify, internally uses [jsonwebtoken](https://github.com/auth0/node-jsonwebtoken).
+JWT utils for Fastify, internally it uses [fast-jwt](https://github.com/nearform/fast-jwt).
+
+**NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.
```
Good catch!

Eomm replied:
Since we are talking about this migration, could we mention a simple upgrade list?

Mainly it concerns the fields name from `issuer` to `iss` etc..

darkgl0w replied:
Good point.
Examples need an updated to reflect those changes

radomird commented:
> LGTM with a little addition to the `README.md` note.
> 
> I spotted this breaking use case in the wild, but I guess this PR will land in a semver major. It's not a blocking case to me as I always use values expressed in milliseconds but this plugin is widely used so WDYT ? Is it worth adding this feature to `fast-jwt` before landing this PR ?

@mcollina @simoneb  what do you think?

simoneb commented:
> > LGTM with a little addition to the `README.md` note.
> > I spotted this breaking use case in the wild, but I guess this PR will land in a semver major. It's not a blocking case to me as I always use values expressed in milliseconds but this plugin is widely used so WDYT ? Is it worth adding this feature to `fast-jwt` before landing this PR ?
> 
> @mcollina @simoneb what do you think?

I'm not sure I understand what's the breaking use case mentioned by @darkgl0w

mcollina commented on jwt.js:
```diff
@@ -186,42 +188,66 @@ function fastifyJwt (fastify, options, next) {
-      options = Object.assign({}, verifyOptions)
-    }
+    const veriferConfig = checkAndMergeVerifyOptions(options, callback)
+    const verifier = createVerifier(veriferConfig.options)
```
This must be moved _outside_ of the `verify` function, otherwise the caching will not take effect. The caching must be documented as well.

mcollina commented on jwt.js:
```diff
@@ -186,42 +188,66 @@ function fastifyJwt (fastify, options, next) {
-    if (typeof callback === 'function') {
-      jwt.sign(payload, secretOrPrivateKey, options, callback)
+    const signerConfig = checkAndMergeSignOptions(options, callback)
+    const signer = createSigner(signerConfig.options)
```
We should not be creating a signer every time. It's expensive and remove most of the benefits of fast-jwt.

mcollina commented on jwt.js:
```diff
@@ -140,7 +141,8 @@ function fastifyJwt (fastify, options, next) {
     }
 
-    return jwt.decode(token, options)
+    const decoder = createDecoder(options)
```
We should not be creating a decoder every time. It's expensive and remove most of the benefits of fast-jwt.

mcollina commented on jwt.js:
```diff
@@ -268,7 +294,10 @@ function fastifyJwt (fastify, options, next) {
       function sign (secretOrPrivateKey, callback) {
-        jwt.sign(payload, secretOrPrivateKey, options.sign || options, callback)
+        const signerOptions = mergeOptionsWithKey(options.sign || options, secretOrPrivateKey)
+        const signer = createSigner(signerOptions)
```
We should not be creating a signer every time. It's expensive and remove most of the benefits of fast-jwt.

mcollina commented on jwt.js:
```diff
@@ -350,21 +379,31 @@ function fastifyJwt (fastify, options, next) {
-          if (err instanceof jwt.TokenExpiredError) {
+        try {
+          const verifierOptions = mergeOptionsWithKey(options.verify || options, secretOrPublicKey)
+          const verifier = createVerifier(verifierOptions)
```
We should not be creating a verifier every time. It's expensive and remove most of the benefits of fast-jwt.

mcollina reviewed (changes requested):
There is still some work needed to integrate this as it's currently not delivering most of the performance benefits of fast-jwt.

mcollina commented:
@ShogunPanda could you review?

darkgl0w commented:
> I'm not sure I understand what's the breaking use case mentioned by @darkgl0w

@simoneb An example use case that will break:

```javascript
  fastify.register(jwt, {
    secret: {
        private: 'private_key_here',
        public: 'public_key_here'
    },
    sign: {
      algorithm: 'RS256',
      // this will fail with this PR because string describing a time span are not converted in ms by `fast-jwt`
      expiresIn: '1 day'
    },
    verify: {
      algorithm: ['RS256']
    }
  });
```

After this PR it must be changed to something like this :
```javascript
  fastify.register(jwt, {
    secret: {
        private: 'private_key_here',
        public: 'public_key_here'
    },
    sign: {
      algorithm: 'RS256',
      expiresIn: 86_400_000 // 1 day in milliseconds
    },
    verify: {
      algorithm: ['RS256']
    }
  });
```

simoneb commented:
> > I'm not sure I understand what's the breaking use case mentioned by @darkgl0w
> 
> @simoneb An example use case that will break:
> 
> ```js
>   fastify.register(require('fastify-jwt'), {
>     secret: {
>         private: 'private_key_here',
>         public: 'public_key_here'
>     },
>     sign: {
>       algorithm: 'RS256',
>       // this will fail with this PR because string describing a time span are not converted to ms by `fast-jwt`
>       expiresIn: '1 day'
>     },
>     verify: {
>       algorithm: ['RS256']
>     }
>   });
> ```
> 
> After this PR it must be changed to something like this :
> 
> ```js
>   fastify.register(require('fastify-jwt'), {
>     secret: {
>         private: 'private_key_here',
>         public: 'public_key_here'
>     },
>     sign: {
>       algorithm: 'RS256',
>       expiresIn: 86_400_000 // 1 day in milliseconds
>     },
>     verify: {
>       algorithm: ['RS256']
>     }
>   });
> ```

Gotcha, makes sense. No strong opinion about whether to do it now or in an upcoming PR.

jsumners commented:
> No strong opinion about whether to do it now or in an upcoming PR.

If this PR isn't a semver major, then this PR should include `ms` to handle the conversion so as not to make it a breaking change.

mcollina commented:
I think we should add support for `ms` in this library. It seems an unneeded breakage.

darkgl0w commented:
> 
> 
> I think we should add support for `ms` in this library. It seems an unneeded breakage.

In this case it could be worth looking at [this fork of `ms`](https://github.com/lukeed/ms) in terms of performance. WDYT ?

ShogunPanda commented:
@mcollina I've reviewed it. I haven't found anything else than what you already mentioned. I do agree that all signers and verifiers must be reused as much as we can. But given that options in fastify-jwt decorators can be changed at any call, this might be problematic. We might have to maintain a map of signers/verifiers/decoders according to the provided options.

About `ms` support, I do also agree that it should be added back. At the end of the day it's just intercepting those options before forwarding them to `fast-jwt`.

mcollina commented:
> > I think we should add support for `ms` in this library. It seems an unneeded breakage.
> 
> In this case it could be worth looking at [this fork of `ms`](https://github.com/lukeed/ms) in terms of performance. WDYT ?

I don't think `ms` should be in the hot path for this module. If it is we have a problem.

radomird commented:
Following the comment of @ShogunPanda :

> @mcollina I've reviewed it. I haven't found anything else than what you already mentioned. I do agree that all signers and verifiers must be reused as much as we can. But given that options in fastify-jwt decorators can be changed at any call, this might be problematic. We might have to maintain a map of signers/verifiers/decoders according to the provided options.

I would suggest to have a separate PR for optimising the solution. I would open a draft PR for that, where we could discuss how to solve the issues that were brought up in this comment. @mcollina what do you think?

mcollina commented:
> I would suggest to have a separate PR for optimising the solution. I would open a separate issue for that, where we could discuss how to solve the issues that were brought up in this comment. @mcollina what do you think?

No. Moving from jsonwebtoken to fast-jwt is _mostly_ about performance. Doing the change without any performance benefits is breakage for our users without much gain.

mcollina commented:
You could do a separate PR that target _this_ PR..

Or if you want we can land this in a `next` branch. Anyhow, this is not hitting "master" without that change.

radomird commented:
Added conversion to milliseconds using the proposed library [@lukeed/ms](https://github.com/lukeed/ms) (proposed by @darkgl0w ). I've tried to minimise the impact on the hot path.

radomird commented:
> You could do a separate PR that target _this_ PR..
> 
> Or if you want we can land this in a `next` branch. Anyhow, this is not hitting "master" without that change.

Clear. Thanks.

climba03003 commented on .github/workflows/ci.yml:
```diff
@@ -16,7 +16,7 @@ jobs:
     strategy:
       matrix:
-        node-version: [10, 12, 14, 16]
+        node-version: [12, 14, 16]
```
Is it a must to drop `node@10`? If yes, I think we can simplify the process and target to a major bump.

mcollina replied:
This is going to be a major bump anyway

radomird replied:
It's a must - with a new release of `fast-jwt` we dropped support for v10 as it uses some functionalities from the newer versions. As it is mentioned earlier it is going to be a major version.

radomird commented:
@mcollina I need your input before I push any changes regarding the optimisation. Should I go for:
- _simple solution_ - check if there are new options provided and then if yes, then create a new instance of signer/decoder/verifier and if not then we use the "global" one
OR
- _complicated solution_ - keep a global options cache and then cache all new options - in order to do this I might introduce a function for hashing the options object in order to keep track of it

I personally would prefer the simpler solution, as in all honesty, the complicated seems like an overkill.

kibertoad commented:
@radomird Aren't there existing memoization libraries that can handle per-parameter combination caching for you?

radomird commented:
> @radomird Aren't there existing memoization libraries that can handle per-parameter combination caching for you?

Indeed there are libraries for that (my preference [fast-memoize.js ](https://github.com/caiogondim/fast-memoize.js)) but I wanted Matteo's opinion of whether we should go that route, as then we include a new dependency to the library.

mcollina commented:
The simple solution described above is ok.

radomird commented:
Simple solution implemented

ShogunPanda commented:
LGTM. Great job!

darkgl0w commented on README.md:
```diff
@@ -411,43 +414,37 @@ fastify.register(require('fastify-jwt'), {
+* `algorithms`: List of strings with the names of the allowed algorithms. By default, all algorithms are accepted.
+* `complete`: Return an object with the decoded header, payload, signature and input (the token part before the signature), instead of just the content of the payload. Default is `false`.
+* `cache`: A positive number specifying the size of the verified tokens cache (using LRU strategy). Setting this to `true` is equivalent to provide the size 1000. When enabled the  performance is dramatically improved. By default the cache is disabled.
+* `cacheTTL`: The maximum time to live of a cache entry (in milliseconds). If the token has a earlier expiration or the verifier has a shorter `maxAge`, the earlier takes precedence. The default is `600000`, which is 10 minutes.
```
```suggestion
* `cacheTTL`: The maximum time to live of a cache entry (in milliseconds). If the token has a earlier expiration or the verifier has a shorter `maxAge`, the earlier takes precedence. The default is `600000`, which is 10 minutes.
* `maxAge`: The maximum allowed age (in seconds) for tokens to still be valid. It is expressed in seconds or a string describing a time span. E.g.: `60`, `"2 days"`, `"10h"`, `"7d"`. A numeric value is interpreted as a seconds count. If you use a string be sure you provide the time units (days, hours, etc.), otherwise milliseconds unit is used by default (`"120"` is equal to `"120ms"`). By default this is not checked.
```

darkgl0w reviewed (approved):
LGTM, good work ^^

I have left some suggestions (mostly documentation related).

darkgl0w commented on README.md:
```diff
@@ -411,43 +413,39 @@ fastify.register(require('fastify-jwt'), {
-* `mutatePayload`: if true, the sign function will modify the payload object directly. This is useful if you need a raw reference to the payload after claims have been applied to it but before it has been encoded into a token.
+* `key`: A string or a buffer containing the secret for `HS*` algorithms or the PEM encoded public key for `RS*`, `PS*`, `ES*` and `EdDSA` algorithms. The key can also be a function accepting a Node style callback or a function returning a promise. If provided, it will override the value of [secret](#secret-required) provided in the options.
+* `algorithm`: The algorithm to use to sign the token. The default is autodetected from the key, using RS256 for RSA private keys, HS256 for plain secrets and the correspondent ES or EdDSA algorithms for EC or Ed* private keys.
+* `mutatePayload`: If the original payload must be modified in place (via Object.assign) and thus will result changed to the caller funciton.
```
```suggestion
* `mutatePayload`: If `true`, the original payload will be modified in place (via `Object.assign`) by the signing function. This is useful if you need a raw reference to the payload after claims have been applied to it but before it has been encoded into a token.
```
Forgot to change this sentence to make it more clear.

darkgl0w replied:
```suggestion
* `mutatePayload`: If `true`, the original payload will be modified in place (via `Object.assign`) by the signing function. This is useful if you need a raw reference to the payload after claims have been applied to it but before it has been encoded into a token.
```

darkgl0w reviewed (commented):
Forgot to change the `mutatePayload` sentence to make it more clear.

radomird commented:
I've ran some manual benchmarks, comparing the old (using `jsonwebtoken`) implementation and the new (using `fast-jwt`). 

I did **only a limited number of runs** (7) and took the average of that and compared. So, **this is by no means a real thorough benchmark** (with 1000's of calls), but it gives an idea of the improvements.

#### OLD (`jsonwebtoken` implementation)
| method | 1 | 2 | 3 | 4 | 5 | 6 | 7 | AVG (not incl. the 1st run) |
| ------ | :---: | :---: | :---: | :---: | :---: | :---: | :---: | ---------------------------: |
| sign | 6.797 | 3.063 | 3.127 | 1.158 | 2.933 | 3.506 | 4.209 | **2.999 ms**
| jwtSign | 7.216 | 3.822 | 3.106 | 4.750 | 4.801 | 2.922 | 3.665 | **3.844 ms**
| verify | 3.077 | 0.347 | 0.380 | 0.380 | 0.403 | 0.536 | 0.424 | **0.411 ms**
| jwtVerify | 1.862 | 0.509 | 0.405 | 0.652 | 0.384 | 0.502 | 0.509 | **0.493 ms**

#### NEW (`fast-jwt` implementation)
| method | 1 | 2 | 3 | 4 | 5 | 6 | 7 | AVG (not incl. the 1st run) |
| ------ | :---: | :---: | :---: | :---: | :---: | :---: | :---: | ---------------------------: |
| sign | 6.499 | 2.130 | 2.106 | 3.147 | 2.145 | 2.039 | 2.162 | **2.288 ms**
| jwtSign | 8.605 | 2.024 | 2.025 | 2.861 | 2.058 | 2.064 | 3.147 | **2.363 ms**
| verify | 1.413 | 0.280 | 0.285 | 0.274 | 0.288 | 0.257 | 0.256 | **0.273 ms**
| jwtVerify | 1.533 | 0.415 | 0.288 | 0.386 | 0.306 | 0.286 | 0.259 | **0.323 ms**

#### Improvements (NEW vs OLD) based on the runs above
| method | improvement |
| ------ | -----------: |
| sign | 23% |
| jwtSign | 38% |
| verify | 33% |
| jwtVerify | 34% |

simoneb commented on jwt.test-d.ts:
```diff
@@ -26,14 +26,14 @@ const secretOptions = {
   sign: {
-    expiresIn: '1h'
+    expiresIn: 36000
   },
   cookie: {
     cookieName: 'jwt',
     signed: false
   },
   verify: {
-    maxAge: '1h',
+    maxAge: 36000,
```
shall we not have tests also with strings?

radomird replied:
We do have them here: 
- https://github.com/fastify/fastify-jwt/blob/3665579076b65f0cdcefe6fd0f797524fdd2e740/test/test.test.js#L2478
- https://github.com/fastify/fastify-jwt/blob/3665579076b65f0cdcefe6fd0f797524fdd2e740/test/test.test.js#L2499

simoneb replied:
these are typings tests, they're two different things :)

darkgl0w replied:
They were not covered in the types and triggered typescript errors, I have added some suggestions in order to make strings describing a time span valid.
However it can be improved by adding a JSDoc example covering string time span use case to make it easier for the end user.

simoneb reviewed (commented):
Overall LGTM, a couple of inline comments (some may be outdated as it took me a while to review this and you may have addressed some of them already).

One main question: shall we write a short migration guide so users know what to change when upgrading?

darkgl0w commented on jwt.test-d.ts:
```diff
@@ -26,14 +26,14 @@ const secretOptions = {
   secret: 'supersecret',
   sign: {
-    expiresIn: '1h'
+    expiresIn: 36000
```
```suggestion
    expiresIn: 3600
```

darkgl0w commented on jwt.test-d.ts:
```diff
@@ -26,14 +26,14 @@ const secretOptions = {
   },
   verify: {
-    maxAge: '1h',
+    maxAge: 36000,
```
```suggestion
    maxAge: '1 hour',
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -126,20 +136,20 @@ declare module 'fastify' {
-    jwtVerify<Decoded extends VerifyPayloadType>(options: jwt.VerifyOptions, callback: VerifyCallback<Decoded>): void
+    jwtVerify<Decoded extends VerifyPayloadType>(callback: VerifierCallback): void
+    jwtVerify<Decoded extends VerifyPayloadType>(options: FastifyJwtVerifyOptions, callback: VerifierCallback): void
+    jwtVerify<Decoded extends VerifyPayloadType>(options?: Partial<VerifierOptions>): Promise<Decoded>
+    jwtVerify<Decoded extends VerifyPayloadType>(callback: VerifierCallback): void
+    jwtVerify<Decoded extends VerifyPayloadType>(options: Partial<VerifierOptions>, callback: VerifierCallback): void
```
```suggestion
    jwtVerify<Decoded extends VerifyPayloadType>(options?: Partial<VerifyOptions>): Promise<Decoded>
    jwtVerify<Decoded extends VerifyPayloadType>(callback: VerifierCallback): void
    jwtVerify<Decoded extends VerifyPayloadType>(options: Partial<VerifyOptions>, callback: VerifierCallback): void
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -126,20 +136,20 @@ declare module 'fastify' {
-    jwtSign(payload: SignPayloadType, options: jwt.SignOptions, callback: jwt.SignCallback): void
+    jwtSign(payload: SignPayloadType, callback: SignerCallback): void
+    jwtSign(payload: SignPayloadType, options: FastifyJwtSignOptions, callback: SignerCallback): void
+    jwtSign(payload: SignPayloadType, options?: Partial<SignerOptions>): Promise<string>
+    jwtSign(payload: SignPayloadType, callback: SignerCallback): void
+    jwtSign(payload: SignPayloadType, options: Partial<SignerOptions>, callback: SignerCallback): void
```
```suggestion
    jwtSign(payload: SignPayloadType, options?: Partial<SignOptions>): Promise<string>
    jwtSign(payload: SignPayloadType, callback: SignerCallback): void
    jwtSign(payload: SignPayloadType, options: Partial<SignOptions>, callback: SignerCallback): void
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
-  decode: jwt.DecodeOptions
-  verify: jwt.VerifyOptions
+  decode: Partial<DecoderOptions>
+  verify: Partial<VerifierOptions>
```
```suggestion
  verify: Partial<VerifyOptions>
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
-  decode: jwt.DecodeOptions
-  verify: jwt.VerifyOptions
+  decode: Partial<DecoderOptions>
+  verify: Partial<VerifierOptions>
```
```suggestion
  verify: Partial<VerifyOptions>
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
 
 export interface FastifyJwtSignOptions {
-  sign?: jwt.SignOptions
+  sign?: Partial<SignerOptions>
```
```suggestion
  sign?: Partial<SignOptions>
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
-  verify<Decoded extends VerifyPayloadType>(token: string, options?: jwt.VerifyOptions): Decoded
-  verify<Decoded extends VerifyPayloadType>(token: string, callback: VerifyCallback<Decoded>): void
-  verify<Decoded extends VerifyPayloadType>(token: string, options: jwt.VerifyOptions, callback: VerifyCallback<Decoded>): void
+  verify<Decoded extends VerifyPayloadType>(token: string, options?: Partial<VerifierOptions>): Decoded
+  verify<Decoded extends VerifyPayloadType>(token: string, callback: VerifierCallback): void
+  verify<Decoded extends VerifyPayloadType>(token: string, options: Partial<VerifierOptions>, callback: VerifierCallback): void
```
```suggestion
  verify<Decoded extends VerifyPayloadType>(token: string, options?: Partial<VerifyOptions>): Decoded
  verify<Decoded extends VerifyPayloadType>(token: string, callback: VerifierCallback): void
  verify<Decoded extends VerifyPayloadType>(token: string, options: Partial<VerifyOptions>, callback: VerifierCallback): void
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
-  sign(payload: SignPayloadType, options?: jwt.SignOptions): string
-  sign(payload: SignPayloadType, callback: jwt.SignCallback): void
-  sign(payload: SignPayloadType, options: jwt.SignOptions, callback: jwt.SignCallback): void
+  sign(payload: SignPayloadType, options?: Partial<SignerOptions>): string
+  sign(payload: SignPayloadType, callback: SignerCallback): void
+  sign(payload: SignPayloadType, options: Partial<SignerOptions>, callback: SignerCallback): void
```
```suggestion
  sign(payload: SignPayloadType, options?: Partial<SignOptions>): string
  sign(payload: SignPayloadType, callback: SignerCallback): void
  sign(payload: SignPayloadType, options: Partial<SignOptions>, callback: SignerCallback): void
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -81,42 +91,42 @@ export interface FastifyJWTOptions {
-    sign: jwt.SignOptions
-    verify: jwt.VerifyOptions
+    decode: Partial<DecoderOptions>
+    sign: Partial<SignerOptions>
+    verify: Partial<VerifierOptions>
```
```suggestion
    sign: Partial<SignOptions>
    verify: Partial<VerifyOptions>
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -37,29 +37,39 @@ export type UserType = FastifyJWT extends { user: infer T }
-  sign?: jwt.SignOptions
-  verify?: jwt.VerifyOptions & { extractToken?: (request: fastify.FastifyRequest) => string | void }
+  decode?: Partial<DecoderOptions>
+  sign?: Partial<SignerOptions>
+  verify?: Partial<VerifierOptions> & { extractToken?: (request: fastify.FastifyRequest) => string | void }
```
```suggestion
  sign?: Partial<SignOptions>
  verify?: Partial<VerifyOptions> & { extractToken?: (request: fastify.FastifyRequest) => string | void }
```

darkgl0w commented on jwt.d.ts:
```diff
@@ -37,29 +37,39 @@ export type UserType = FastifyJWT extends { user: infer T }
 export interface DecodeCallback<Decoded extends DecodePayloadType> {
   (err: Error, decoded: Decoded): void
 }
 
```
```suggestion

export interface SignOptions extends Omit<SignerOptions, "expiresIn" | "notBefore"> {
  expiresIn: number | string;
  notBefore: number | string;
}

export interface VerifyOptions extends Omit<VerifierOptions, "maxAge"> {
  maxAge: number | string;
}

```

radomird commented on README.md:
```diff
@@ -5,7 +5,10 @@
+JWT utils for Fastify, internally it uses [fast-jwt](https://github.com/nearform/fast-jwt).
+
+**NOTE:** The plugin has been migrated from using `jsonwebtoken` to `fast-jwt`. Even though `fast-jwt` has 1:1 feature implementation with `jsonwebtoken`, some _exotic_ implementations might break. In that case please open an issue with details of your implementation.
+In addition, note that if you made use of strings describing a time span [vercel/ms](https://github.com/vercel/ms) (eg: `"1h"`, `"10 days"`, `"1 month"`...), you **must** convert all these references to **milliseconds**.
```
This has been updated. The additional note was removed in https://github.com/fastify/fastify-jwt/pull/184/commits/3665579076b65f0cdcefe6fd0f797524fdd2e740.

mcollina commented on README.md:
```diff
@@ -411,43 +413,39 @@ fastify.register(require('fastify-jwt'), {
+* `key`: A string or a buffer containing the secret for `HS*` algorithms or the PEM encoded public key for `RS*`, `PS*`, `ES*` and `EdDSA` algorithms. The key can also be a function accepting a Node style callback or a function returning a promise. If provided, it will override the value of [secret](#secret-required) provided in the options.
+* `algorithms`: List of strings with the names of the allowed algorithms. By default, all algorithms are accepted.
+* `complete`: Return an object with the decoded header, payload, signature and input (the token part before the signature), instead of just the content of the payload. Default is `false`.
+* `cache`: A positive number specifying the size of the verified tokens cache (using LRU strategy). Setting this to `true` is equivalent to provide the size 1000. When enabled the  performance is dramatically improved. By default the cache is disabled.
```
I would go ahead and enable the cache by default.

Wdyt?

darkgl0w replied:
Unless it is enabled by default in `fast-jwt` I would not make it `true` by default in `fastify-jwt`.

But we can properly document this feature (with an example use case and maybe a link to [this NearForm article](https://www.nearform.com/blog/improve-json-web-tokens-performance-in-node-js/)) and highlight the benefits in the `README.md` and in the upcoming package changelog.

radomird replied:
I agree with @darkgl0w

radomird commented on jwt.d.ts:
```diff
@@ -37,29 +37,39 @@ export type UserType = FastifyJWT extends { user: infer T }
-export type TokenOrHeader = jwt.JwtHeader | { header: jwt.JwtHeader; payload: any }
+// standard names https://www.rfc-editor.org/rfc/rfc7515.html#section-4.1
+// same interface as in jsonwebtoken lib - https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/jsonwebtoken/index.d.ts
+export interface JwtHeader {
```
No we don't. Maybe it does make sense to move it there. I would make another ticket for this as it requires (another) release of `fast-jwt`.

simoneb replied:
Yes let's do that!

radomird replied:
Added:
- https://github.com/nearform/fast-jwt/issues/147
- https://github.com/fastify/fastify-jwt/issues/188

darkgl0w reviewed (commented):
Added some changes suggestions to types. (mostly string time span related)

mcollina reviewed (approved):
lgtm

mcollina commented:
You might want to run some benchmarks using wrk or autocannon.

radomird commented:
I think we are good to go with merging. After the workflow is approved and the tests are green I'll go ahead and merge this PR.

Eomm reviewed (approved):
LGTM, just a suggestion on the readme

simoneb commented:
> LGTM, just a suggestion on the readme

Yes absolutely, I suggested the same. We should have a section in the readme or other sensible place and link it from the release notes

radomird commented:
@Eomm @simoneb & @darkgl0w I've updated the README with breaking changes and also all of the examples in the README have been updated to use the new option names.

darkgl0w commented on README.md:
```diff
@@ -713,6 +713,21 @@ fastify.get('/', async (request, replay) => {
+  - `audience` should be changed to `allowedAud`
+  - `issuer` should be changed to `allowedIss`
+  - `subject` should be changed to `allowedSub`
+
```
```suggestion
  - `jwtid` should be changed to `allowedJti`
  - `nonce` should be changed to `allowedNonce`

```

radomird replied:
Good catch

darkgl0w commented on README.md:
````diff
@@ -713,6 +713,21 @@ fastify.get('/', async (request, replay) => {
 
 ```
 
+## `jsonwebtoken` migration changes
````
We should use something like: 'Upgrading from `fastify-jwt@3`'
WDYT ?

radomird replied:
Yeah that could be also - I have no strong preference. Maybe others can say something as well?
@simoneb  @Eomm

simoneb replied:
yes let's specify which version we're migrating from. note that this could also live in a standalone doc instead of cluttering the main readme

radomird replied:
Done

darkgl0w commented on README.md:
```diff
@@ -187,16 +187,16 @@ fastify.register(jwt, {
   sign: {
     algorithm: 'ES256',
-    issuer: 'api.example.tld'
+    isss: 'api.example.tld'
```
```suggestion
    iss: 'api.example.tld'
```

radomird replied:
Whoops held my finger too long on the keyboard 😸

darkgl0w reviewed (commented):
Seems good to me ^^
I have left a few nits

darkgl0w commented on UPGRADING.md:
```diff
@@ -0,0 +1,17 @@
+- **verify** options:
+  - `audience` should be changed to `allowedAud`
+  - `issuer` should be changed to `allowedIss`
+  - `subject` should be changed to `allowedSub`
```
```suggestion
  - `subject` should be changed to `allowedSub`
  - `jwtid` should be changed to `allowedJti`
  - `nonce` should be changed to `allowedNonce`

```

darkgl0w reviewed (commented):
That's way better with a separate file :+1:

mcollina reviewed (commented):
Could you add a bit of a migration guide at the top of the README?

There are a few API changes here

simoneb commented:
> Could you add a bit of a migration guide at the top of the README?
> 
> There are a few API changes here

There's an UPGRADING.md doc added in this PR

mcollina commented on UPGRADING.md:
```diff
@@ -0,0 +1,17 @@
+
+### Upgrading from 3.x to 4.0
+
+In `v4` we migrated away from using `jsonwebtoken` to `fast-jwt`. This introduced the following breaking changes:
```
If the names of the properties are the only thing remaining.. could we avoid them entirely? We could mske fast-jwt accept both.

darkgl0w replied:
There is also a breaking change in the `decode` options (`json` entry has been removed and `checkTyp` has been introduced).
This PR will land as a major and apart from those changes and maybe some __exotic__ use cases the migration should be flawless and TBH I highly prefer `fast-jwt` options naming as it is way more obvious.
Accepting both can be confusing and can lead to unexpected behaviours (e.g.: `issuer` and `allowedIss` are both supported in verify options, if by mistake I register something like `{ verify: { issuer: 'one', allowedIss: 'two' } }` which one will be considered valid ?).
In the actual state it's just a matter of searching and replacing a few option names and ts typings will help to spot which ones needs to be updated.

radomird replied:
I agree with @darkgl0w

radomird commented:
Folks, I think the PR is, finally, ready to be merged. If some small issues arise after the release, we will address them separately.

mcollina reviewed (approved):
lgtm

radomird commented:
Can anyone with `write access` merge the PR? I don't have sufficient access level to do this 😄 
@mcollina

climba03003 reviewed (approved):
No blocking for me. LGTM.

Uzlopak commented:
Neat!

I assume that @lukeed will be happy to read that his package will be used in a this fastify-plugin. (Pinging him on purpose ;) )

mcollina reviewed (approved):
lgtm

lukeed commented:
Nice! :) thanks for the ping
